### ISSUE TITLE
TypeError: can't convert undefined to object on schema with nested definitions

### ISSUE THREAD

**Issue by pgillet** [mozilla-services/react-jsonschema-form#140]
Hi,

I get the following error: `TypeError: can't convert undefined to object`

With this type of schema which contains nested definitions with $ref:

```json
{
  "title": "Parent obj",
  "properties": {
    "name": {
      "title": "Name",
      "readOnly": false,
      "type": "string"
    },
    "childObj": {
      "title": "Child object",
      "readOnly": false,
      "type": "object",
      "$ref": "#/definitions/childObj"
    }
  },
  "definitions": {
    "childObj": {
      "type": "object",
      "properties": {
        "otherName": {
          "title": "Other name",
          "readOnly": false,
          "type": "string"
        }
      }
    }
  },
  "type": "object",
  "$schema": "http://json-schema.org/draft-04/schema#"
}
```

The error occurs in `utils#computeDefaults` method.

As far as I understand:

1. The defaults are first computed for the referenced schema (`childObj` in my case) within the last block of the method after the comment `// We need to recur for object schema inner default values`. 
2. Then, we come back in the calling block (`if ("$ref" in schema)`). At this point, the schema variable is the `childObj` property in the upper schema:

```json
{
  "title": "Child object",
  "readOnly": false,
  "type": "object",
  "$ref": "#/definitions/childObj"
}
```
 
3. Finally, as the schema type is equal to `object`, we end up in the last block again, where the call to  `schema.properties` throws the error. 

Does it make sense to you?

Thanks for your feedback.

Pascal

**Comment by n1k0** [comment]
In my testing, problem seems goes away when you don't define a type along `$ref` usage:

Before (doesn't work)

```json
{
  "title": "Parent obj",
  "properties": {
    "name": {
      "type": "string"
    },
    "childObj": {
      "$ref": "#/definitions/childObj"
    }
  },
  "definitions": {
    "childObj": {
      "type": "object",
      "properties": {
        "otherName": {
          "type": "string"
        }
      }
    }
  },
  "type": "object"
}
```

After (works)
 
```json
{
  "title": "Parent obj",
  "properties": {
    "name": {
      "type": "string"
    },
    "childObj": {
      "type": "object",
      "$ref": "#/definitions/childObj"
    }
  },
  "definitions": {
    "childObj": {
      "type": "object",
      "properties": {
        "otherName": {
          "type": "string"
        }
      }
    }
  },
  "type": "object"
}
```

**Comment by pgillet** [comment]
Thank you for your quick answer.
Yes, it works fine if the type along the $ref is removed. 

But my schema is automatically generated by Jackson on a domain Java object with nested object properties. I am not sure that the JSON schema standard requires to specify the type for nested definitions. I think it is necessarily an "object" and it should be implicit here...

Anyway, I saw your fix. Thank you.

**Comment by n1k0** [comment]
I'm reopening as the fix has not been reviewed nor landed just yet :)

Thanks for your prompt feedback.

**Comment by sadasiva8598** [comment]
@n1k0 n1k0 closed this in #141 on 18 Apr 2016
n1k0 added a commit that referenced this issue on 18 Apr 2016
@n1k0
Fixes #140: Schema definition ref should override local schema type. (#…
691197c
@sadasiva8598
